### DISCUSSION TITLE
Lost type information when returning from a function.

### DISCUSSION THREAD

**Discussion by mzalevski** [colinhacks/zod#369] [Q&A] [answered]
Hi, I have a problem with lost type information when returning from a function.

The `manageData` function returns `data` and `valid`.

If `valid` is `true` it means that the safeParse is successful and the `data` type should be `S`.
If `valid` is `false` it means that the safeParse is not successful and the `data` type should be `I`.

It seems that the types from within the function are recognized correctly but after returning from it the info is lost (photos below).

Is there something I can do or is it a TypeScript thing?

```
import * as z from "zod";

interface User {
  firstName: string;
  lastName: string;
}

const userSchema: z.ZodSchema<User> = z.object({
  firstName: z.string(),
  lastName: z.string(),
});

interface Params<S, I> {
  schema: z.ZodType<S, z.ZodTypeDef, S>;
  initialData: I;
}

const manageData = <S, I>({ schema, initialData }: Params<S, I>) => {
  let data = initialData;
  const setData = (newData: any) => (data = newData);

  const parseResult = schema.safeParse(data);

  if (parseResult.success) {
    return { data: parseResult.data, valid: true, setData }; // data is recognized as `S` type -> correct!
  } else {
    return { data: data, valid: false, setData }; // data is recognized as `I` type -> correct!
  }
};

const { data, valid } = manageData({
  schema: userSchema,
  initialData: { firstName: "John" },
});

if (valid) {
  console.log(data); // data is recognized as S | I union type -> I want to have `S` here
} else {
  console.log(data); // data is recognized as S | I union type -> I want to have `I` here
}

```

**it seems that the types from within the function are recognized correctly `(property) data: S` and `(property) data: I`**

![image](https://user-images.githubusercontent.com/43313417/112746273-0e2f6280-8fae-11eb-8749-a7d12818f985.png)
![image](https://user-images.githubusercontent.com/43313417/112746282-1ab3bb00-8fae-11eb-9fbf-911f2b235af3.png)

**but the types after returning from the function are merged into an union**

![image](https://user-images.githubusercontent.com/43313417/112746294-2acb9a80-8fae-11eb-98bc-a80e8178750b.png)
![image](https://user-images.githubusercontent.com/43313417/112746310-446ce200-8fae-11eb-9493-0f0d92fbac8e.png)

**Comment by colinhacks** [comment, marked as answer]
For discriminated unions like this, the discriminator (in this case "valid") needs to be a literal (like `"adsf"` or `true`, not `string` or `boolean`. You need to tell TypeScript to treat literal values as literals. In this case just add `as const` to the boolean literals like this, and it works as expected:

```ts
  if (parseResult.success) {
    return { data: parseResult.data, valid: true as const, setData }; // data is recognized as `S` type -> correct!
  } else {
    return { data: data, valid: false as const, setData }; // data is recognized as `I` type -> correct!
  }
```

**Reply by mzalevski** [reply]
Thanks! I did as you wrote and from what I see, it does work, but only if you do not destructure the returned object. Any ideas how to make destructuring work?
```
const manager = manageData({
  schema: userSchema,
  initialData: { firstName: "John" },
});

if (manager.valid) {
  console.log(manager.data); // data is recognized as `S` type -> correct!
} else {
  console.log(manager.data); // data is recognized as `I` type -> correct!
}
```
I'd love to destructure the object like this, but the type info is lost.
```

const { data, valid } = manageData({
  schema: userSchema,
  initialData: { firstName: "John" },
});

if (valid) {
  console.log(data); // still an S | I union
} else {
  console.log(data); // still an S | I union
}
```

**Reply by mzalevski** [reply]
It seems to be a bigger problem: https://github.com/microsoft/TypeScript/issues/12184 & https://stackoverflow.com/questions/59786104/how-to-make-discriminated-unions-work-with-destructuring. Nevertheless, maybe you have some ideas :/

**Reply by colinhacks** [reply]
Yeah, you can't destructure with discriminated unions, and that's unlikely to change.
